src/software/ai/hl/stp/action/kick_action.h

```diff
@@ -15,7 +15,7 @@ class KickAction : public Action
     /**
      * Creates a new KickAction
      */
-    explicit KickAction();
+    explicit KickAction() = delete;
```

Comment:
Does anyone know if the `explicit` keyword do anything when the function is being deleted? @jonathanlew @akhilveeraghanta

Reply:
It's redundant https://stackoverflow.com/questions/2836939/purpose-of-explicit-default-constructors

Reply:
Just to confirm, when a constructor is declared with the `explicit` keyword then it is not required to have a separate `delete` constructor line since the `explicit` keyword prevents a default constructor from being created?
So I would only need to `delete` the constructors which which are not `explicit`? @jonathanlew

Reply:
So for example, for `explicit CherryPickTactic(const World& world, const Rectangle& target_region);` it is not required to explicitly remove the constructor (`CherryPickTactic() = delete;`), however I should for readability? 
Similarly for [example](https://github.com/UBC-Thunderbots/Software/blob/cb6700fbc7d419e4c0a124eb3c21a7d7a481438d/src/software/ai/hl/stp/tactic/mutable_tactic_visitor.h#L34) for `virtual ~MutableTacticVisitor() = default;` it is not necessary to have a `delete` constructor line, because `virtual` destructor will automatically prevent a default constructor from being created?

From reading [this article](https://docs.microsoft.com/en-us/cpp/cpp/explicitly-defaulted-and-deleted-functions?view=msvc-160) my understanding is that it is not necessary to explicitly remove `explicit` constructors and `virtual` destructors. I would appreciate it if you could clarify this as I'm a bit confused.

Reply:
A few things: 

1. `virtual ~MutableTacticVisitor() = default` is needed because `MutableTacticVisitor` is a virtual class, so it's unrelated to deleting constructors. Also, `~[class name]()` is a destructor declaration, and we're not concerned with misused destructors because we don't _usually_ use raw pointers

2. **Note that explicitly deleting a default constructor doesn't mean using the keyword `explicit` (you can if you'd like), it's supposed to mean _not implicitly_.** Deleting default constructors we don't use offers this benefit in the error message

implicitly deleted function:
```
/usr/include/c++/7/ext/new_allocator.h:136:4: error: **no matching** function for call to 'CherryPickTactic::CherryPickTactic()'
  { ::new((void *)__p) _Up(std::forward<_Args>(__args)...); }
    ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```
_not implicitly_ deleted function:
```
/usr/include/c++/7/ext/new_allocator.h:136:4: error: use of **deleted** function 'CherryPickTactic::CherryPickTactic()'
  { ::new((void *)__p) _Up(std::forward<_Args>(__args)...); }
    ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

The deleted function makes it clear that it's not a mismatch of arguments with the class, but that instead the class expressly deleted that constructor.